src/RateLimits.sol

```diff
@@ -0,0 +1,137 @@
+    /**********************************************************************************************/
+
+    constructor(
+        address admin_
+    ) {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin_);
+    }
+
+    /**********************************************************************************************/
+    /*** Admin functions                                                                        ***/
+    /**********************************************************************************************/
+
+    function setRateLimit(
+        bytes32 key,
+        uint256 maxAmount,
+        uint256 slope,
+        uint256 lastAmount,
+        uint256 lastUpdated
+    )
+        public override onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        require(lastAmount <= maxAmount,        "RateLimits/invalid-lastAmount");
+        require(lastUpdated <= block.timestamp, "RateLimits/invalid-lastUpdated");
+
+        _data[key] = RateLimitData({
+            maxAmount:   maxAmount,
+            slope:       slope,
+            lastAmount:  lastAmount,
+            lastUpdated: lastUpdated
+        });
+
+        emit RateLimitSet(key, maxAmount, slope, lastAmount, lastUpdated);
+    }
+
+    function setRateLimit(
+        bytes32 key,
+        uint256 maxAmount,
+        uint256 slope
+    )
+        external override
+    {
+        setRateLimit(key, maxAmount, slope, 0, block.timestamp);
+    }
+
+    function setUnlimitedRateLimit(
+        bytes32 key
+    )
+        external override
+    {
+        setRateLimit(key, type(uint256).max, 0, 0, block.timestamp);
+    }
+
+    /**********************************************************************************************/
+    /*** Getter Functions                                                                       ***/
+    /**********************************************************************************************/
+
+    function getData(bytes32 key) external override view returns (RateLimitData memory) {
+        return _data[key];
+    }
+
+    function getCurrentRateLimit(bytes32 key) public override view returns (uint256) {
+        RateLimitData memory d = _data[key];
+
+        // Unlimited rate limit case
+        if (d.maxAmount == type(uint256).max) {
+            return type(uint256).max;
+        }
+
+        return _min(
+            d.slope * (block.timestamp - d.lastUpdated) + d.lastAmount,
+            d.maxAmount
+        );
+    }
+
+    /**********************************************************************************************/
+    /*** Controller functions                                                                   ***/
+    /**********************************************************************************************/
+
+    function triggerRateLimit(bytes32 key, uint256 amountToDecrease)
+        external override onlyRole(CONTROLLER) returns (uint256 newLimit)
+    {
+        require(amountToDecrease > 0, "RateLimits/invalid-amountToDecrease");
```

Comment:
This is a fair point. I actually wanted to prevent the case where the limit was not set up. I'd rather have it error if `maxAmount` is zero (AKA it's not activated), but your point about having the zero amount not fail makes sense. I've updated the code.